pkg/server/admin.go

```diff
@@ -1397,6 +1397,43 @@ func (s *adminServer) checkReadinessForHealthCheck() error {
 		return status.Errorf(codes.Unavailable, "node is shutting down")
 	}
 
+	if !s.server.sqlServer.acceptingClients.Get() {
+		return status.Errorf(codes.Unavailable, "node is not accepting SQL clients")
+	}
+
+	// Also check the availability of the SQL service.
```

Comment:
CC @chrisseto to weigh in!

In CC, we use `/health?ready=1` as a readiness probe (https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/). This probe is used in two places (I *think* just these two places):

1. When doing upgrades, we roll a single node over, then wait until all nodes in the cluster are ready, meaning the probe passes on all nodes. Then we continue to another node.
2. If a node starts failing the readiness probe, k8s will stop the node from "receiving traffic through Kubernetes Services" (https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#define-readiness-probes). It's actually not clear to me what this means with a stateful protocol like SQL; are existing connections dropped on the floor? I wonder if @chrisseto knows.

Re: 1, querying these system tables seems like a useful thing to do. We will halt bad upgrades earlier.

Re: 2, I am not so sure. All these system tables become critical dependencies. If one of em starts having problems, then our k8s setup may lead to a big drop in capacity, as all nodes will detect this at the same time. This may make the actual problem worse. The feedback loop introduced worries me.

What do folks think about this? There is no fundamental reason that we need to use the same probe for 1 & 2 btw. OTOH, k8s may make certain assumptions that make it somewhat finicky not to. Again @chrisseto to comment on that.

Reply:
> it's actually not clear to me what this means with a stateful protocol like SQL; are existing connections dropped on the floor?

Existing connections will **not** be dropped. The pod will just not receive any new connections from the load balancer. Though it gets a bit weird, as other nodes will still be able to discover this one (our service discover Service doesn't require readiness). So it would be possible for a node to be executing dist sql queries (Can someone fact check me here?).

> There is no fundamental reason that we need to use the same probe for 1 & 2 btw. OTOH, k8s may make certain assumptions that make it somewhat finicky not to.

Can you say more about this? AFAIK they have to be the same probe and they should be the same probe from k8s perspective. Do you mean that we could have a manual probe for the rolling upgrade?

My main question is whether or not this check adds a dependency on other nodes being live/ready. If I have an 8 node cluster and only 3 nodes are up, will all nodes become unready?

Our Statefulsets are setup to manage all pods in parallel and nodes will be able to connect to unready nodes, so I don't think this will have any unintended consequences.

Reply:
> Existing connections will not be dropped.

Very interesting!